File: examples/typescript/glucose_monitor.ts
/**
 * CGM glucose monitor overlay — draggable desktop panel showing blood glucose from Nightscout.
 *
 * Connects to a Nightscout instance when NIGHTSCOUT_URL is set, otherwise
 * runs in simulated mode with random-walk glucose values.
 *
 * Setup:
 *   cd examples/typescript
 *   npm install winpane   (or: npm link ../../../bindings/node)
 *
 * Usage:
 *   npx tsx glucose_monitor.ts
 *
 * Environment variables (optional):
 *   NIGHTSCOUT_URL   — base URL of your Nightscout site
 *   NIGHTSCOUT_TOKEN — API token for authenticated access
 */
// ── winpane design tokens ──────────────────────────────────────
// Surface base: #121216  Glass: +e4  Solid: +ff  Muted: +f2
// Elevated:     #1c1c21  Interactive: #26262cff  Hover: #303038ff
// Border:       #ffffff12  Text:      #e8e8edff   Muted: #9494a0ff
// Accent:       #528bffff  Success:   #34d399ff   Warning: #fbbf24ff
// Danger:       #ef4444ff  Radius: 10/6 px
// ────────────────────────────────────────────────────────────────
import { WinPane } from "winpane";

// ── Types ──────────────────────────────────────────────────────

interface GlucoseReading {
  sgv: number;
  direction: string;
  timestamp: number; // Date.now() ms
}

// ── Helper functions ───────────────────────────────────────────

function directionToArrow(direction: string): string {
  const arrows: Record<string, string> = {
    DoubleUp: "⇈",
    SingleUp: "↑",
    FortyFiveUp: "↗",
    Flat: "→",
    FortyFiveDown: "↘",
    SingleDown: "↓",
    DoubleDown: "⇊",
  };
  return arrows[direction] ?? "?";
}

function bgColorForSgv(sgv: number): string {
  if (sgv >= 70 && sgv <= 180) return "#12281ee4"; // green-tinted
  if (sgv >= 181 && sgv <= 250) return "#282412e4"; // amber-tinted
  return "#281212e4"; // red-tinted (<70 or >250)
}

// Staleness measures time since fetch, not CGM reading time.
function stalenessText(timestampMs: number): { text: string; color: string } {
  const elapsed = Date.now() - timestampMs;
  const text = elapsed < 60000 ? "just now" : `${Math.floor(elapsed / 60000)} min ago`;
  const color = elapsed > 900000 ? "#ef4444ff" : "#9494a0cc";
  return { text, color };
}

async function fetchNightscout(url: string, token?: string): Promise<GlucoseReading | null> {
  try {
    let endpoint = `${url}/api/v1/entries/current.json`;
    if (token) endpoint += `?token=${token}`;
    const response = await fetch(endpoint);
    const data = await response.json();
    const entry = data[0];
    return {
      sgv: entry.sgv,
      direction: entry.direction,
      timestamp: Date.now(),
    };
  } catch {
    return null;
  }
}

function simulateReading(prevSgv: number): GlucoseReading {
  const delta = Math.trunc(Math.random() * 31 - 15);
  const sgv = Math.max(40, Math.min(350, prevSgv + delta));

  let direction: string;
  if (delta > 10) direction = "SingleUp";
  else if (delta > 5) direction = "FortyFiveUp";
  else if (delta > -5) direction = "Flat";
  else if (delta > -10) direction = "FortyFiveDown";
  else direction = "SingleDown";

  return { sgv, direction, timestamp: Date.now() };
}

// ── Main ───────────────────────────────────────────────────────

const wp = new WinPane();
const panel = wp.createPanel({ width: 140, height: 93, x: 1760, y: 902, draggable: true, dragHeight: 28 });
wp.setCaptureExcluded(panel, true);

// Initial bg
wp.setRect(panel, "bg", {
  x: 0, y: 0, width: 140, height: 93,
  fill: bgColorForSgv(120),
  cornerRadius: 10,
  borderColor: "#ffffff12",
  borderWidth: 1,
});
wp.show(panel);

// Title bar in drag region
wp.setRect(panel, "title_bg", {
  x: 0, y: 0, width: 140, height: 28,
  fill: "#1c1c21ff",
});
wp.setText(panel, "title", {
  text: "Glucose",
  x: 8, y: 6,
  fontSize: 13,
  bold: true,
  color: "#9494a0ff",
});

const nightscoutUrl = process.env.NIGHTSCOUT_URL;
const nightscoutToken = process.env.NIGHTSCOUT_TOKEN;
const pollInterval = nightscoutUrl ? 5 * 60 * 1000 : 30 * 1000;

if (nightscoutUrl) {
  console.log("winpane glucose_monitor: polling Nightscout every 5 min.");
} else {
  console.log("winpane glucose_monitor: simulated mode (set NIGHTSCOUT_URL for live data).");
}
console.log("Press Ctrl+C to exit.");

let lastPoll = 0; // force immediate first poll
let currentReading: GlucoseReading = { sgv: 120, direction: "Flat", timestamp: Date.now() };

setInterval(async () => {
  const now = Date.now();
  if (now - lastPoll >= pollInterval) {
    if (nightscoutUrl) {
      const reading = await fetchNightscout(nightscoutUrl, nightscoutToken);
      if (reading) currentReading = reading;
    } else {
      currentReading = simulateReading(currentReading.sgv);
    }
    lastPoll = Date.now();
  }

  // Update bg
  wp.setRect(panel, "bg", {
    x: 0, y: 0, width: 140, height: 93,
    fill: bgColorForSgv(currentReading.sgv),
    cornerRadius: 10,
    borderColor: "#ffffff12",
    borderWidth: 1,
  });

  // Update reading
  const arrow = directionToArrow(currentReading.direction);
  wp.setText(panel, "reading", {
    text: `${currentReading.sgv} ${arrow}`,
    x: 12, y: 34,
    fontSize: 30,
    fontFamily: "Consolas",
    bold: true,
    color: "#e8e8edff",
  });

  // Update staleness
  const stale = stalenessText(currentReading.timestamp);
  wp.setText(panel, "staleness", {
    text: stale.text,
    x: 12, y: 70,
    fontSize: 12,
    color: stale.color,
  });
}, 1000);

// Graceful cleanup
process.on("SIGINT", () => {
  wp.destroy(panel);
  wp.close();
  process.exit(0);
});

File: examples/python/glucose_monitor.py
"""
CGM glucose monitor overlay — draggable desktop panel showing blood glucose from Nightscout.

Connects to a Nightscout instance when NIGHTSCOUT_URL is set, otherwise
runs in simulated mode with random-walk glucose values. Stdlib only — no
pip dependencies.

Usage:
    python glucose_monitor.py

Requires winpane-host.exe in PATH or specify the path below.

Environment variables (optional):
    NIGHTSCOUT_URL   — base URL of your Nightscout site
    NIGHTSCOUT_TOKEN — API token for authenticated access

Note: This simple send/receive pattern works because the overlay has no
interactive elements beyond the native drag handle (no event polling needed).
"""
# ── winpane design tokens ──────────────────────────────────────
# Surface base: #121216  Glass: +e4  Solid: +ff  Muted: +f2
# Elevated:     #1c1c21  Interactive: #26262cff  Hover: #303038ff
# Border:       #ffffff12  Text:      #e8e8edff   Muted: #9494a0ff
# Accent:       #528bffff  Success:   #34d399ff   Warning: #fbbf24ff
# Danger:       #ef4444ff  Radius: 10/6 px
# ────────────────────────────────────────────────────────────────
import subprocess
import json
import time
import os
import random
import urllib.request
import urllib.parse

HOST_PATH = "winpane-host"

DIRECTION_ARROWS = {
    "DoubleUp": "⇈",
    "SingleUp": "↑",
    "FortyFiveUp": "↗",
    "Flat": "→",
    "FortyFiveDown": "↘",
    "SingleDown": "↓",
    "DoubleDown": "⇊",
}


def direction_to_arrow(direction):
    return DIRECTION_ARROWS.get(direction, "?")


def bg_color_for_sgv(sgv):
    if 70 <= sgv <= 180:
        return "#12281ee4"
    if 181 <= sgv <= 250:
        return "#282412e4"
    return "#281212e4"


# Staleness measures time since fetch, not CGM reading time.
def staleness_text(reading_time):
    elapsed = time.time() - reading_time
    if elapsed < 60:
        text = "just now"
    else:
        text = f"{int(elapsed // 60)} min ago"
    color = "#ef4444ff" if elapsed > 900 else "#9494a0cc"
    return text, color


def fetch_nightscout(url, token=None):
    try:
        endpoint = f"{url}/api/v1/entries/current.json"
        if token:
            endpoint += f"?token={urllib.parse.quote(token, safe='')}"
        req = urllib.request.Request(endpoint)
        with urllib.request.urlopen(req, timeout=10) as resp:
            data = json.loads(resp.read().decode())
        entry = data[0]
        return {
            "sgv": int(entry["sgv"]),
            "direction": entry.get("direction", "NONE"),
            "timestamp": time.time(),
        }
    except Exception:
        return None


def simulate_reading(prev_sgv):
    delta = random.randint(-15, 15)
    sgv = max(40, min(350, prev_sgv + delta))
    if delta > 10:
        direction = "SingleUp"
    elif delta > 5:
        direction = "FortyFiveUp"
    elif delta > -5:
        direction = "Flat"
    elif delta > -10:
        direction = "FortyFiveDown"
    else:
        direction = "SingleDown"
    return {"sgv": sgv, "direction": direction, "timestamp": time.time()}


def main():
    proc = subprocess.Popen(
        [HOST_PATH],
        stdin=subprocess.PIPE,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        text=True,
        bufsize=1,
    )

    request_id = 0

    def send(method, params=None):
        nonlocal request_id
        request_id += 1
        msg = {
            "jsonrpc": "2.0",
            "method": method,
            "params": params or {},
            "id": request_id,
        }
        line = json.dumps(msg)
        proc.stdin.write(line + "\n")
        proc.stdin.flush()
        resp_line = proc.stdout.readline()
        return json.loads(resp_line)

    try:
        resp = send("create_panel", {"width": 140, "height": 93, "x": 1760, "y": 902, "draggable": True, "drag_height": 28})
        sid = resp["result"]["surface_id"]
        print(f"Created panel: {sid}")

        send("set_capture_excluded", {"surface_id": sid, "excluded": True})

        # Initial bg
        send("set_rect", {
            "surface_id": sid,
            "key": "bg",
            "x": 0, "y": 0, "width": 140, "height": 93,
            "fill": bg_color_for_sgv(120),
            "corner_radius": 10,
            "border_color": "#ffffff12",
            "border_width": 1,
        })

        # Title bar in drag region
        send("set_rect", {
            "surface_id": sid,
            "key": "title_bg",
            "x": 0, "y": 0, "width": 140, "height": 28,
            "fill": "#1c1c21ff",
            "corner_radius": 10,
        })
        send("set_text", {
            "surface_id": sid,
            "key": "title",
            "text": "Glucose",
            "x": 8, "y": 6,
            "font_size": 13,
            "bold": True,
            "color": "#9494a0ff",
        })

        send("show", {"surface_id": sid})

        nightscout_url = os.environ.get("NIGHTSCOUT_URL")
        nightscout_token = os.environ.get("NIGHTSCOUT_TOKEN")
        poll_interval = 5 * 60 if nightscout_url else 30

        if nightscout_url:
            print("winpane glucose_monitor: polling Nightscout every 5 min.")
        else:
            print("winpane glucose_monitor: simulated mode (set NIGHTSCOUT_URL for live data).")
        print("Press Ctrl+C to exit.")

        last_poll = 0.0  # force immediate first poll
        current_reading = {"sgv": 120, "direction": "Flat", "timestamp": time.time()}

        while True:
            now = time.time()
            if now - last_poll >= poll_interval:
                if nightscout_url:
                    reading = fetch_nightscout(nightscout_url, nightscout_token)
                    if reading:
                        current_reading = reading
                else:
                    current_reading = simulate_reading(current_reading["sgv"])
                last_poll = time.time()

            # Update bg
            send("set_rect", {
                "surface_id": sid,
                "key": "bg",
                "x": 0, "y": 0, "width": 140, "height": 93,
                "fill": bg_color_for_sgv(current_reading["sgv"]),
                "corner_radius": 10,
                "border_color": "#ffffff12",
                "border_width": 1,
            })

            # Update reading
            arrow = direction_to_arrow(current_reading["direction"])
            send("set_text", {
                "surface_id": sid,
                "key": "reading",
                "text": f"{current_reading['sgv']} {arrow}",
                "x": 12, "y": 34,
                "font_size": 30,
                "font_family": "Consolas",
                "bold": True,
                "color": "#e8e8edff",
            })

            # Update staleness
            stale_text, stale_color = staleness_text(current_reading["timestamp"])
            send("set_text", {
                "surface_id": sid,
                "key": "staleness",
                "text": stale_text,
                "x": 12, "y": 70,
                "font_size": 12,
                "color": stale_color,
            })

            time.sleep(1)

    except KeyboardInterrupt:
        print("\nShutting down...")
    finally:
        if proc.poll() is None:
            try:
                send("destroy", {"surface_id": sid})
            except (BrokenPipeError, OSError):
                pass
        try:
            proc.stdin.close()
        except (BrokenPipeError, OSError):
            pass
        proc.wait(timeout=5)


if __name__ == "__main__":
    main()
